### PR TITLE
Feat custom typecast

### LINKED ISSUE #36: Expose decode configuration

### PR DESCRIPTION
**Pull request by 634750802**
close #36

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -23,15 +23,59 @@ export interface Config {
...
+  TIMESTAMP = 'TIMESTAMP',
+  BIT = 'BIT',
+  JSON = 'JSON'
+}
+
+export type DecodeConfig = Partial<
+  {
+    [P in ColumnType]: (rawValue: string) => any
+  } & {
+    [key: string]: (rawValue: string) => any
```
Comment (shiyuhang0) [inline src/config.ts:79]: why we need `& {[key: string]: (rawValue: string) => any}`

File: __tests__/decode.test.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+describe('decode', () => {
+  describe('cast', () => {
+    test('use default decode method if config not provided', () => {
+      expect(cast(makeField(ColumnType.BIGINT), '21412421321421521321', {})).toEqual('21412421321421521321')
+      expect(cast(makeField(ColumnType.FLOAT), '1.2', {})).toEqual(1.2)
+      expect(cast(makeField(ColumnType.INT), '1', {})).toEqual(1)
+    })
+
+    test('override default decode method if config not provided', () => {
```
Comment (shiyuhang0) [inline __tests__/decode.test.ts:13]:
```suggestion
    test('override default decode method if config provided', () => {
```

File: src/config.ts [hunk 3]
```diff
@@ -23,15 +23,59 @@ export interface Config {
...
+  LONGBLOB = 'LONGBLOB',
+  DATE = 'DATE',
+  TIME = 'TIME',
+  DATETIME = 'DATETIME',
+  TIMESTAMP = 'TIMESTAMP',
+  BIT = 'BIT',
+  JSON = 'JSON'
+}
+
+export type DecodeConfig = Partial<
```
Comment (shiyuhang0) [inline src/config.ts:75]: I prefer `  [key in ColumnType]?: (rawValue: string) => any` rather than Partial

Comment (shiyuhang0) [inline src/config.ts:75]: DecodeConfig => Decode? or do you have a better name?

File: src/config.ts (hunk 1 above)
Comment (634750802) [reply src/config.ts:79]: for future unlisted field types (like `VECTOR`)

File: src/config.ts (hunk 3 above)
Comment (634750802) [reply src/config.ts:75]: maybe Decoders?

File: src/config.ts (hunk 1 above)
Comment (shiyuhang0) [reply src/config.ts]: I think 
```
{
  [P in ColumnType]?: (rawValue: string) => any
} & {
  [key: string]: (rawValue: string) => any
}
```
is equal to 

```
  key: string]: (rawValue: string) => any
```

Both of them can't limit the type of key in typescript. I prefer to use one of `[P in ColumnType]?: (rawValue: string) => any` and `[key: string]: (rawValue: string) => any`

File: src/config.ts (hunk 3 above)
Comment (shiyuhang0) [reply src/config.ts]: Decoders looks better than DecodeConfig. Do you have any suggestion @zhangyangyu ?

File: src/config.ts (hunk 1 above)
Comment (shiyuhang0) [reply src/config.ts]: The type system in typescript aims to check code in compile time.
With `[P in ColumnType]`, we will meet the compile error with the key not included in ColumnType. However `[key: string]` breaks this. This loses the meaning of defining ColumnType.

If we want to support other types. Just add them in ColumnType. (By the way, I think you can also pass any other value in js)

Comment (634750802) [reply src/config.ts]:
```ts
{
  [P in ColumnType]?: (rawValue: string) => any
} & {
  [key: string]: (rawValue: string) => any
}
```

This one supports both IDE auto completion (DX) and compiler check for other types, but it's OK to not support custom column types (maybe never needed).

I'll keep just the first half.

**Review by shiyuhang0 — APPROVED** [review]

**Pull request merged**